Hardware Keyboard: Linux (GTK)

dkwingsmt opened the pull request:
## Description

This PR contains the Linux (GTK) changes for the Hardware Keyboard project.

Classes has been adjusted as follows:

<table>
  <tr>
    <th>Before</th>
    <th>After</th>
    <th>Explain</th>
  </tr>
  <tr>
    <td rowspan="2">FlKeyEventPlugin</td>
    <td>FlKeyboardManager</td>
    <td>Handles redispatching and detects redispatched events.</td>
  </tr>
  <tr>
    <td>FlKeyChannelResponder</td>
    <td>Sends raw events through method channel and collect responses.</td>
  </tr>
  <tr>
    <td> - </td>
    <td>FlKeyEmbedderResponder</td>
    <td>Converts raw events to FlutterKeyEvent, sends them through embedder API and collect responses.</td>
  </tr>
</table>


## Related Issues

- Umbrella: flutter/flutter#44918

## Tests

I added the following tests:

- A number of tests for `FlKeyboardManager`
- A number of tests for `FlKeyEmbedderResponder`

Migrated tests of `FlKeyEventPlugin` to `FlKeyChannelResponder`.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [ ] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation.
- [ ] All existing and new tests are passing.
- [ ] I am willing to follow-up on review comments in a timely manner.


## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.


## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

gspencergoog commented on shell/platform/linux/fl_key_embedder_responder.cc:
```diff
@@ -0,0 +1,719 @@
+      return;
+    }
+
+    const int standard_current_stage =
```
couldn't this just be:
```cpp
standard_current_stage = current_stage % kNumStages;
```

dkwingsmt reviewed (commented):
I addressed the comments, and made significant improvement to the synthesization logic to support remapped keys, as well as some tests.

dkwingsmt commented:
@gspencergoog Finally made all tests passed. I've refactored the code to using a new class `FlKeyEvent` that wraps the system class `GdkEvent` (the same way as we discussed about iOS).

dkwingsmt commented:
@gspencergoog Can you take another look? I've addressed the issues, and I've refactored the code to using a Flutter event class that wraps the information of the system event class, just as how we discussed for iOS. The build should also be green now.

gspencergoog commented on shell/platform/linux/fl_key_channel_responder.cc:
```diff
@@ -0,0 +1,282 @@
+static void fl_key_channel_responder_init(FlKeyChannelResponder* self) {}
+
+// Creates a new FlKeyChannelResponder instance, with a messenger used to send
+// messages to the framework, an FlTextInputPlugin used to handle key events
```
```suggestion
// messages to the framework, and an FlTextInputPlugin that is used to handle key events
```

gspencergoog commented on shell/platform/linux/fl_key_channel_responder.cc:
```diff
@@ -0,0 +1,293 @@
+  //
+  // This code has the flaw that if a key event is missed due to the app losing
+  // focus, then this state will still think the key is pressed when it isn't,
+  // but that is no worse than for "regular" keys until we implement the
```
Yes, makes sense now: of course this is the old one that doesn't do that.

gspencergoog commented on shell/platform/linux/fl_key_channel_responder.cc:
```diff
@@ -0,0 +1,282 @@
+  // as being "pressed" whenever the lock is on, which breaks a lot of
+  // interactions (for example, if shift-lock is on, tab traversal is broken).
+  //
+  // TODO(gspencergoog): get rid of this tracked state when we are tracking the
```
You can remove this TODO, since you now do that in the new code, and we're not planning on updating this code to do that.

gspencergoog commented on shell/platform/linux/fl_key_channel_responder_test.cc:
```diff
@@ -0,0 +1,204 @@
+static FlKeyEvent _g_key_event;
+}  // namespace
+
+// Clone #string onto the heap.
```
Don't add markup in implementation comments.
```suggestion
// Clone string onto the heap.
```

gspencergoog commented on shell/platform/linux/fl_keyboard_manager.cc:
```diff
@@ -0,0 +1,412 @@
+         (static_cast<uint64_t>(event->is_press ? GDK_KEY_PRESS
+                                                : GDK_KEY_RELEASE) &
+          0xffff)
+             << 32 |
```
This has some weird indenting.  Maybe it would make sense to split into some const values and | those together for readability? (I know this is just copied from the previous implementation, but it's ugly. :-))

dkwingsmt replied:
Sure.

Speaking of this, this logic is a bit weird to me. I think this hash should _not_ include time, and _should_ include keyval. What was your idea?

gspencergoog replied:
The idea was to identify the exact event from GTK, so the time is crucial to that. If you didn't include time, then a rapidly pressed key might mean that it matches a different down event for that key, and handle them out of order. These three attributes were enough to make the hash be unique for the fields we cared about.  You could add in keyval, or even substitute it for keycode if you want: I don't think that would make much difference.

gspencergoog commented on shell/platform/linux/fl_keyboard_manager.cc:
```diff
@@ -0,0 +1,412 @@
+    g_return_if_fail(removed == pending);
+    bool should_redispatch = false;
+    if (!pending->any_handled) {
+      // If no responders have handled, send it to text plugin.
```
```suggestion
      // If no responders have been handled, send it to text plugin.
```

gspencergoog commented on shell/platform/linux/fl_keyboard_manager.cc:
```diff
@@ -0,0 +1,412 @@
+      if (self->text_input_plugin == nullptr ||
+          !fl_text_input_plugin_filter_keypress(self->text_input_plugin,
+                                                pending->event)) {
+        // If text plugin doesn't handle either, redispatch.
```
```suggestion
        // If the text plugin doesn't handle it either, redispatch.
```

gspencergoog commented on shell/platform/linux/key_mapping.cc:
```diff
@@ -0,0 +1,467 @@
+                             uint64_to_gpointer(fl_key));
+}
+
+void initialize_xkb_to_physical_key(GHashTable* table) {
```
I know this is using gtk, but why not use a std::map here for all of these tables?  You can initialize them statically that way, and won't need to call an initializer function.

dkwingsmt replied:
I'm inclining to use full C (instead of C++) in the Linux embedding. I doubt if it's necessary, but I think that's we're committing to since we've been using GTK class (instead of C++ class) all over.

gspencergoog replied:
I don't think you need to do that: there's already C++ in the linux embedding, because it uses the embedder API. I agree that in general we should try to use GTK classes for the GTK embedding, but in this case, it means we can't statically define these data structures, and there's no need to waste startup time setting them up just to use a C/GTK interface.

gspencergoog replied:
I'm actually running into something similar in the iOS implementation: because the KeyCodes.mm file uses NSDictionary, those structures are not always initialized when loaded in XCode as a dylib, and it crashes when in the debugger. So I'm converting them to use std::map, which has no such issue.

dkwingsmt replied:
The embedder API is pure C AFAIK, otherwise we wouldn't need to use pure function and user_data instead of the convenient lambdas.

gspencergoog replied:
Yes, the API is C-compatible, but the implementation is C++.

dkwingsmt replied:
Still, why do we keep using GTK class everywhere then? Almost all classes can be converted to C++ classes since they won't be passed to GTK APIs anyway, and we'll have so much cleaner code.

gspencergoog replied:
How would using a C++ implementation here restrict users of the Linux embedding? Isn't this just an implementation detail of the key handling?

gspencergoog replied:
> Still, why do we keep using GTK class everywhere then? 

As far as I know, it is only a style thing: when in GTK land, do as GTK does, that kind of thing. I think the idea is that someone maintaining the GTK code might be more familiar with GTK style than some hybrid (and it would be a hybrid, since we'd have to transform to GTK structs somewhere). Style shouldn't slow things down unnecessarily, though, IMO.

dkwingsmt replied:
I'll change the map to std::map, but seriously, the tediousness of GTK classes is beyond acceptance to the point that I'm missing Objective C classes. We might want to reconsider this decision some day.

gspencergoog replied:
Yes, I agree. I almost think that we could abstract a lot of this code into some common C++ classes that take delegates for the platform specific work.

gspencergoog replied:
I say "almost" because I'm pretty sure there are some snags in that plan. But there is a lot of similar logic in the implementations now.

dkwingsmt replied:
Keyboard manager is the most likely class to be extracted, but then we have iOS. Embedder responder is much less likely: the maps are not shared, the synthesization logic is very different, and the rest is mostly trivial.

dkwingsmt commented:
@gspencergoog Can you take another look? 
For your convenience, this is the diff after your last review. The changes by this PR are everything under the linux folder (including `fl_settings_plugin.cc`).
https://github.com/flutter/engine/compare/3ecc134bc68f96b01e2dd3fd921e4c4406db454f..e967132b5c39bb444d31c8d7d6a79acf390c28d2

gspencergoog commented:
@dkwingsmt I think that link compares the wrong revisions, unless you've been VERY busy :-) (it currently shows ~5000 lines of difference in 115 files for me). No worries, though, I'll just go through the code.

dkwingsmt commented:
@gspencergoog That's because I've updated with master since your last review, and that's what I mean by "the changes by this PR are everything under the linux folder". :)

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

To be honest, I have looked this over a number of times, but there's so much code here, it's hard to be sure that things are all "correct". In general, it looks good though.

dkwingsmt commented:
@gspencergoog Thanks for the review!
